File: day1/part1.py
from sys import stdin
from typing import List

def parse(input: str) -> List[str]:
    return input.splitlines()

def solve(lines: List[str]) -> int:
    count = 0
    dial = 50
    for line in lines:
        op, n = line[0], int(line[1:])
        if op == 'L': n = -n
        dial += n
        dial %= 100
        if dial == 0:
            count += 1
    return count

if __name__ == "__main__":
    print(solve(parse(stdin.read())))

File: day1/part1.hs
parse :: String -> [String]
parse = lines

solve :: [String] -> Int
solve = count 0 . turns

count :: Eq a => a -> [a] -> Int
count x = length . filter (== x)

turns :: [String] -> [Int]
turns = scanl turn dial where
    dial = 50
    turn n = (`mod` 100) . (n +) . lineToTurn

lineToTurn :: String -> Int
lineToTurn line =
    let op = head line
        n = read $ tail line
    in  if op == 'L'
        then -n
        else n

main :: IO ()
main = interact (show . solve . parse)
